File: src/Canon.Emit/PostgresEmitter.fs
namespace Canon.Emit.Postgres

open System
open Canon.Emit
open Canon.Introspect

/// Emits Postgres DDL (CREATE TABLE) from the domain schema.
type PostgresEmitter() =
    
    interface IEmitter with
        member this.Emit(tables: TableDef list) =
            let sb = Text.StringBuilder()
            
            for table in tables do
                sb.AppendLine($"CREATE TABLE {table.Name} (") |> ignore
                
                let colDefs = 
                    table.Columns |> List.map (fun col ->
                        let nullability = if col.IsNullable then "NULL" else "NOT NULL"
                        let typeStr = 
                            match col.MaxLength with
                            | Some len -> $"{col.DataType}({len})"
                            | None -> col.DataType
                            
                        let checks = 
                            if col.CheckConstraints.IsEmpty then ""
                            else 
                                let combinedChecks = String.Join(" AND ", col.CheckConstraints)
                                $" CHECK ({combinedChecks})"
                                
                        $"    {col.Name} {typeStr} {nullability}{checks}"
                    )
                
                let combinedCols = String.Join(",\n", colDefs)
                sb.AppendLine(combinedCols) |> ignore
                sb.AppendLine(");") |> ignore
                sb.AppendLine() |> ignore
                
            [ sb.ToString(), Canon.Core.Fidelity.Exact ] // Return single DDL string for all tables, Postgres is exact to the TableDef

File: playground/src/canonflow/Canon.Emit/PostgresEmitter.js

import { StringBuilder__AppendLine, StringBuilder__AppendLine_Z721C83C5, StringBuilder_$ctor } from "../fable_modules/fable-library-js.5.6.0/System.Text.js";
import { disposeSafe, getEnumerator } from "../fable_modules/fable-library-js.5.6.0/Util.js";
import { join, concat } from "../fable_modules/fable-library-js.5.6.0/String.js";
import { singleton, isEmpty, map } from "../fable_modules/fable-library-js.5.6.0/List.js";
import { toString } from "../fable_modules/fable-library-js.5.6.0/Types.js";
import { Fidelity } from "../Canon.Core/Lineage.js";
import { Emitter } from "./Emitter.js";

export function createEmitter() {
    return new Emitter((tables) => {
        const sb = StringBuilder_$ctor();
        const enumerator = getEnumerator(tables);
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const table = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
                StringBuilder__AppendLine_Z721C83C5(sb, concat("CREATE TABLE ", table.Name, " ("));
                StringBuilder__AppendLine_Z721C83C5(sb, join(",\n", map((col) => {
                    let matchValue;
                    const nullability = col.IsNullable ? "NULL" : "NOT NULL";
                    return `    ${col.Name} ${(matchValue = col.MaxLength, (matchValue == null) ? col.DataType : (`${col.DataType}(${matchValue})`))} ${nullability}${isEmpty(col.CheckConstraints) ? "" : concat(" CHECK (", join(" AND ", col.CheckConstraints), ")")}`;
                }, table.Columns)));
                StringBuilder__AppendLine_Z721C83C5(sb, ");");
                StringBuilder__AppendLine(sb);
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        return singleton([toString(sb), new Fidelity(0, [])]);
    });
}

